From 6a74e2666c521e9c58598ea149693027f22f6159 Mon Sep 17 00:00:00 2001
From: Marko Lindqvist <cazfi74@gmail.com>
Date: Sat, 1 Apr 2023 14:57:06 +0300
Subject: [PATCH 30/30] AI: Fix check if new building enables/disables actions

Check if an action is possible now, and the check if
action will be possible with a new building, were
not equivalent, causing their results to differ when
they should have not.

Changes made to the with-building check:
- Stop passing the improvement to be built as parameter
  to every requirement check. The requirement is likely
  about some other building
- Stop making also real check for any requirements
  that the building in question is present

Originally reported by alain_bkr

See osdn #42169

Signed-off-by: Marko Lindqvist <cazfi74@gmail.com>
---
 ai/default/daicity.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/ai/default/daicity.c b/ai/default/daicity.c
index c698efc70a..527e2625f6 100644
--- a/ai/default/daicity.c
+++ b/ai/default/daicity.c
@@ -1593,11 +1593,17 @@ static void adjust_improvement_wants_by_effects(struct ai_type *ait,
     .city = pcity,
     .building = pimprove,
   };
+
+  /* Do NOT pass building here, as the action might be about
+   * targeting some completely different building, AND
+   * the check to see if the action is possible before
+   * the building is there is also ignoring the buildings.
+   * We don't want those two results to differ for
+   * an unrelated reason to what we are evaluating. */
   const struct req_context actenabler_ctxt = {
     .player = pplayer,
     .city = pcity,
     .tile = city_tile(pcity),
-    .building = pimprove,
   };
 
   /* Remove team members from the equation */
@@ -1785,9 +1791,7 @@ static void adjust_improvement_wants_by_effects(struct ai_type *ait,
             active = FALSE;
             break;
           }
-        }
-
-        if (!is_req_active(&actenabler_ctxt, NULL, preq, RPT_POSSIBLE)) {
+        } else if (!is_req_active(&actenabler_ctxt, NULL, preq, RPT_POSSIBLE)) {
           active = FALSE;
           break;
         }
-- 
2.39.2

